pallets/phala/src/compute/wrapped_balances.rs

```diff
@@ -133,6 +152,248 @@ pub mod pallet {
+			_who: &T::AccountId,
+			_value: Self::Balance,
+		) -> (Self::NegativeImbalance, Self::Balance) {
+			unimplemented!()
+		}
+
+		fn deposit_into_existing(
+			_who: &T::AccountId,
+			_value: Self::Balance,
+		) -> Result<Self::PositiveImbalance, DispatchError> {
+			unimplemented!()
+		}
+
+		fn deposit_creating(_who: &T::AccountId, _value: Self::Balance) -> Self::PositiveImbalance {
+			unimplemented!()
+		}
+
+		fn withdraw(
+			_who: &T::AccountId,
+			_value: Self::Balance,
+			_reasons: WithdrawReasons,
+			_liveness: ExistenceRequirement,
+		) -> Result<Self::NegativeImbalance, DispatchError> {
+			unimplemented!()
+		}
+
+		fn make_free_balance_be(
+			_who: &T::AccountId,
+			_value: Self::Balance,
+		) -> SignedImbalance<Self::Balance, Self::PositiveImbalance> {
+			unimplemented!()
+		}
+	}
+
+	impl<T: Config> ReservableCurrency<T::AccountId> for Pallet<T>
+	where
+		BalanceOf<T>: sp_runtime::traits::AtLeast32BitUnsigned + Copy + FixedPointConvert + Display,
+		T: pallet_assets::Config<AssetId = u32, Balance = BalanceOf<T>>,
+		T: Config + pallet_balances::Config + vault::Config + pallet_elections_phragmen::Config,
+		T: pallet_balances::Config<Balance = BalanceOf<T>>,
+	{
+		fn can_reserve(who: &T::AccountId, value: Self::Balance) -> bool {
+			Self::total_balance(who) >= value
+		}
+		fn reserved_balance(who: &T::AccountId) -> Self::Balance {
+			ElectionReserves::<T>::get(who)
+		}
+		fn reserve(who: &T::AccountId, value: Self::Balance) -> DispatchResult {
+			let actual = value.min(Self::total_balance(who));
+			ElectionReserves::<T>::mutate(who, |reserve| {
+				*reserve += actual;
+			});
+			Ok(())
+		}
+		fn unreserve(who: &T::AccountId, value: Self::Balance) -> Self::Balance {
+			let actual = value.min(ElectionReserves::<T>::get(who));
+			ElectionReserves::<T>::mutate(who, |reserve| {
+				*reserve -= actual;
+			});
+			value - actual
+		}
+		fn slash_reserved(
+			who: &T::AccountId,
+			value: Self::Balance,
+		) -> (Self::NegativeImbalance, Self::Balance) {
+			if value == Zero::zero() {
+				return (NegativeImbalanceOf::<T>::zero(), Zero::zero());
+			}
+			let mut actual = value.min(ElectionReserves::<T>::get(who));
+			let free_wpha: BalanceOf<T> =
+			<pallet_assets::pallet::Pallet<T> as Inspect<T::AccountId>>::balance(
+				T::WPhaAssetId::get(),
+				who,
+			);
+			if free_wpha < actual {
+				let new_debt = actual - free_wpha;
+				actual = free_wpha;
+				SlashDebts::<T>::mutate(who, |debt| {
+					*debt += new_debt;
+				});
+			}
+			ElectionReserves::<T>::mutate(who, |reserve| {
+				*reserve -= actual;
+			});
+			Self::burn_from(who, actual)
+				.expect("burn nft when slash reserved should success; qed.");
+			let imbalance = <T as PhalaConfig>::Currency::withdraw(
+				&T::WrappedBalancesAccountId::get(),
+				actual,
+				WithdrawReasons::all(),
+				ExistenceRequirement::AllowDeath,
+			)
+			.expect("slash imbalance should success; qed.");
+			Self::deposit_event(Event::<T>::ReserveSlashed{
```

Comment:
Please add a field to reflect the debt like: `ReserveSlashed { user, slash, debt }`